### PR TITLE
Epic/front end notifications request model

### LINKED ISSUE #159: Create a notification component (main menu item)
**What:** Create a notification component that is an icon - similar to our current top menu icons.
This icon lights up red to show that a new notification is received. Clicking this icon links to our notification page. 

**How:** Start with Figma design, then create a React component. This may depend on back-end work.

**Why:** Feature enhancement

**Things to consider:**

### LINKED ISSUE #160: Create a notifications page
What: Notification page is a central place for all notifications to be responded to (connection request, messages?).
- This page can make a GET request to the API to fetch relevant notifications
- Notification items are rendered on this page and can be responded to here
How: 
1. Create a figma design.
2. Create a notifications page.

### LINKED ISSUE #161: Create a notification item component
What: An individual notification component to be potentially displayed in a list of other notifications. 
Notification Item contains url link to the particular resource (ie. profile (for connection requests)), a thread, or a DM

### PR DESCRIPTION
**Pull request by davideastmond**
Resolves #159 Resolves #160 Resolves #161

PR for notifications and connection request model.
- Implemented socket client
- Added notifications slice to store
- Implemented notification status icon
- tweak socket connection issues.
- Create notification page
- Connection Request model and functionality
- to be tested in tandum with the API branch `epic/notifications-requests-model`

### CODE REVIEW

File: app/src/components/followButton/FollowButton.tsx [hunk 1]
```diff
@@ -22,30 +24,55 @@ function FollowButton({
...
+                handleCancelAddConnection();
+              },
             },
-          },
-        }}
-      >
-        <span className="Follow-button__text">+ Follow</span>
-      </OptionsMenu>
+          }}
+        >
```
Comment (davideastmond) [inline app/src/components/followButton/FollowButton.tsx]: This works now, but I'm thinking a better approach would be to replace the "Follow" button for the user with the "cancel" option.

**Review by lily-law — APPROVED** [review]
:cool: this is looking good... 
Maybe the socketIO stuff would be better in App for now and the session kept in a ref and with the callback functions in single use useEffects, i.e...
```
/* ...snip... */
function App() {
  const socket: any = useRef();
  const isLoggedIn = useSelector(selectIsLoggedIn, shallowEqual);
  const status = useSelector(selectAuthStatus, shallowEqual);
  const dispatch = useDispatch();
  const userInfo = useSelector(selectUserById("me"), shallowEqual);

  useEffect(() => {
    if (userInfo) {
      socket.current = io("http://localhost:7000");
    }
  }, [userInfo]);

  useEffect(() => {
    if (socket.current) {
      socket.current.on("connect", () => {
        console.log("IO connected!");
        socket.current.emit("myId", userInfo.id);
        console.log("sending my info", userInfo.id);
      });
      socket.current.on("notification", (data:any) => {
        console.log("*** NOTIFICATION FROM SERVER ***", data);
        dispatch(getNotificationsAsync());
      });
    }
  }, [socket.current]);

  // On first load check if authed
/* ...snip... */
```

**Comment by davideastmond** [comment]
> 🆒 this is looking good...
> Maybe the socketIO stuff would be better in App for now and the session kept in a ref and with the callback functions in single use useEffects, i.e...
> 
> ```
> /* ...snip... */
> function App() {
>   const socket: any = useRef();
>   const isLoggedIn = useSelector(selectIsLoggedIn, shallowEqual);
>   const status = useSelector(selectAuthStatus, shallowEqual);
>   const dispatch = useDispatch();
>   const userInfo = useSelector(selectUserById("me"), shallowEqual);
> 
>   useEffect(() => {
>     if (userInfo) {
>       socket.current = io("http://localhost:7000");
>     }
>   }, [userInfo]);
> 
>   useEffect(() => {
>     if (socket.current) {
>       socket.current.on("connect", () => {
>         console.log("IO connected!");
>         socket.current.emit("myId", userInfo.id);
>         console.log("sending my info", userInfo.id);
>       });
>       socket.current.on("notification", (data:any) => {
>         console.log("*** NOTIFICATION FROM SERVER ***", data);
>         dispatch(getNotificationsAsync());
>       });
>     }
>   }, [socket.current]);
> 
>   // On first load check if authed
> /* ...snip... */
> ```

ahh yes, useRef! great idea.

**Review by lily-law — APPROVED** [review]
Very :cool: to have notification working and connection requests.

A couple of bits we need to look at but I'd suggest merging this in and we can make new tickets for these and I can help. 

Styling needs some tweaks, things like the menu in the header "my network" buttons text is squished so its on two lines. I think maybe I could include this and some other bits in a general styling touch up issue.

It would be good to have a way to remove connections

**Comment by davideastmond** [comment]
> Very 🆒 to have notification working and connection requests.
> 
> A couple of bits we need to look at but I'd suggest merging this in and we can make new tickets for these and I can help.
> 
> Styling needs some tweaks, things like the menu in the header "my network" buttons text is squished so its on two lines. I think maybe I could include this and some other bits in a general styling touch up issue.
> 
> It would be good to have a way to remove connections

- Agreed re: styling
hmm! I was pretty sure removing connections functionality still works?

**Comment by lily-law** [comment]
> hmm! I was pretty sure removing connections functionality still works?

I wasn't able to see a button/interface on the front end on the network page or the connection's profile page

**Comment by davideastmond** [comment]
> > hmm! I was pretty sure removing connections functionality still works?
> 
> I wasn't able to see a button/interface on the front end on the network page or the connection's profile page

That's strange. I am able to see it under `my network` - maybe you have to refresh the page?:
image

**Comment by lily-law** [comment]
> > > hmm! I was pretty sure removing connections functionality still works?
> 
> > 
> 
> > I wasn't able to see a button/interface on the front end on the network page or the connection's profile page
> 
> 
> 
> That's strange. I am able to see it under `my network` - maybe you have to refresh the page?:
> 
> image
> 
> 

Okay cool, the functionality is there, must just be a bug, will try later if I can reproduce and will make a new ticket 😄

**Pull request merged**